client/my-sites/email/paths.ts

```diff
@@ -0,0 +1,189 @@
+import { stringify } from 'qs';
+import { isUnderDomainManagementAll, domainManagementRoot } from 'calypso/my-sites/domains/paths';
+
+type QueryStringParameters = { [ key: string ]: string | undefined };
+
+type EmailPathUtilityFunction = (
+	siteName: string | null | undefined,
+	domainName?: string | null,
+	relativeTo?: string | null,
+	urlParameters?: QueryStringParameters
+) => string;
+
+export const emailManagementPrefix = '/email';
+export const emailManagementAllSitesPrefix = '/email/all';
+
+export function isUnderEmailManagementAll( path?: string | null ) {
+	return path?.startsWith( emailManagementAllSitesPrefix + '/' );
+}
+
+// Builds a URL query string from an object. Handles null values.
+const buildQueryString = ( parameters?: QueryStringParameters ) =>
+	parameters ? stringify( parameters, { addQueryPrefix: true, skipNulls: true } ) : '';
+
+function resolveRootPath( relativeTo?: string | null ) {
+	if ( relativeTo === emailManagementAllSitesPrefix || relativeTo === domainManagementRoot() ) {
+		return emailManagementAllSitesPrefix;
+	}
+
+	if ( isUnderEmailManagementAll( relativeTo ) || isUnderDomainManagementAll( relativeTo ) ) {
+		return emailManagementAllSitesPrefix;
+	}
+
+	return emailManagementPrefix;
+}
+
+function getPath(
+	siteName: string | null | undefined,
+	domainName: string | null | undefined,
```

Comment:
Just out of curiosity - have you tried to use `domainName?: string | null`?
To avoid such cases https://github.com/Automattic/wp-calypso/pull/86096/files#diff-4560dae1096ab816ab653ff380644c91fdb17156af430f7fb23ed4097d3a3a26R145 with providing obviously `null`.

Reply:
My reasoning here was that I _didn't_ want to make `domainName` an [optional parameter](https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters). It's fine if a `null` or `undefined` value is given (we know this means that the state is loading), but if `domainName` isn't given at all, we can't construct the full path.

Reply:
The line you link to is interesting, though, because it does illustrate a case where the `domainName` parameter could be left out and the returned path would still be complete.

And looking closer at my own code, it turns out that I've actually already made `domainName` optional for the exported function 😆 (see https://github.com/Automattic/wp-calypso/pull/86096/files#diff-ec526e3657cdfa459d5cfc158337c91ddff11e3113defdfa4afc67b35ed5c399R8)

I'll tweak the function call you referenced!